streamer/src/nonblocking/qos.rs

```diff
@@ -52,3 +52,25 @@ pub(crate) trait QosController<C: ConnectionContext> {
         connection: Connection,
     ) -> impl Future<Output = usize> + Send;
 }
+
+/// Marker trait to indicate what is the shared state for connections
+pub trait ConnectionTableSharedState: Send + Sync {
+    fn as_any_arc(self: Arc<Self>) -> Arc<dyn Any + Send + Sync>;
+}
+
+pub fn get_shared_state<T: ConnectionTableSharedState + 'static>(
+    x: Arc<dyn ConnectionTableSharedState>,
+) -> Option<Arc<T>> {
+    let any_arc = x.as_any_arc();
+    Arc::downcast::<T>(any_arc).ok()
+}
+
+#[cfg(test)]
+pub struct NullConnectionTableSharedState {}
+
+#[cfg(test)]
+impl ConnectionTableSharedState for NullConnectionTableSharedState {
+    fn as_any_arc(self: Arc<Self>) -> Arc<dyn Any + Send + Sync> {
```

Comment:
The impl here is the same for all implementors, but making a default impl did not work out due to compiler not understanding what I want from it...